executor/write.go

```diff
@@ -181,1277 +173,14 @@ func updateRecord(ctx sessionctx.Context, h int64, oldData, newData []types.Datu
-		if err != nil {
-			return nil, errors.Errorf("INSERT INTO %s: %s", e.Table.Meta().Name.O, err)
-		}
-	} else {
-		// If e.Columns are empty, use all columns instead.
-		cols = tableCols
-	}
-	for _, col := range cols {
-		if col.Name.L == model.ExtraHandleName.L {
-			e.hasExtraHandle = true
-			break
-		}
-	}
-
-	// Check column whether is specified only once.
-	err = table.CheckOnce(cols)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-
-	return cols, nil
-}
-
-func (e *InsertValues) lazilyInitColDefaultValBuf() (ok bool) {
-	if e.colDefaultVals != nil {
-		return true
-	}
-
-	// only if values count of insert statement is more than one, use colDefaultVals to store
-	// casted default values has benefits.
-	if len(e.Lists) > 1 {
-		e.colDefaultVals = make([]defaultVal, len(e.Table.Cols()))
-		return true
-	}
-
-	return false
-}
-
-func (e *InsertValues) fillValueList() error {
-	if len(e.Setlist) > 0 {
-		if len(e.Lists) > 0 {
-			return errors.Errorf("INSERT INTO %s: set type should not use values", e.Table)
-		}
-		l := make([]expression.Expression, 0, len(e.Setlist))
-		for _, v := range e.Setlist {
-			l = append(l, v.Expr)
-		}
-		e.Lists = append(e.Lists, l)
-	}
-	return nil
-}
-
-func (e *InsertValues) checkValueCount(insertValueCount, valueCount, genColsCount, num int, cols []*table.Column) error {
-	// TODO: This check should be done in plan builder.
-	if insertValueCount != valueCount {
-		// "insert into t values (), ()" is valid.
-		// "insert into t values (), (1)" is not valid.
-		// "insert into t values (1), ()" is not valid.
-		// "insert into t values (1,2), (1)" is not valid.
-		// So the value count must be same for all insert list.
-		return ErrWrongValueCountOnRow.GenByArgs(num + 1)
-	}
-	if valueCount == 0 && len(e.Columns) > 0 {
-		// "insert into t (c1) values ()" is not valid.
-		return ErrWrongValueCountOnRow.GenByArgs(num + 1)
-	} else if valueCount > 0 {
-		explicitSetLen := 0
-		if len(e.Columns) != 0 {
-			explicitSetLen = len(e.Columns)
-		} else {
-			explicitSetLen = len(e.Setlist)
-		}
-		if explicitSetLen > 0 && valueCount+genColsCount != len(cols) {
-			return ErrWrongValueCountOnRow.GenByArgs(num + 1)
-		} else if explicitSetLen == 0 && valueCount != len(cols) {
-			return ErrWrongValueCountOnRow.GenByArgs(num + 1)
-		}
-	}
-	return nil
-}
-
-func (e *InsertValues) getRows(cols []*table.Column) (rows [][]types.Datum, err error) {
-	// process `insert|replace ... set x=y...`
-	if err = e.fillValueList(); err != nil {
-		return nil, errors.Trace(err)
-	}
-
-	rows = make([][]types.Datum, len(e.Lists))
-	length := len(e.Lists[0])
-	for i, list := range e.Lists {
-		if err = e.checkValueCount(length, len(list), len(e.GenColumns), i, cols); err != nil {
-			return nil, errors.Trace(err)
-		}
-		e.rowCount = uint64(i)
-		rows[i], err = e.getRow(cols, list, i)
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
-	}
-	return
-}
+const (
```

Comment:
This could be merged with in`sert_common.go`.

Reply:
Unfortunately, `DirtyTableAddRow` is used by `write.go` now.